Find possible callee candidates for function pointers and select among those.

fodinabor opened the pull request:
This (currently) has a few restrictions:
- Attributor has to find possible callees in the first place.
- Only function pointers being handed directly to the function under investigation are handled (e.g. `as_cast` is therefore unsupported atm)

So we will want to steer away from direct arg handling (and likely/maybe also from callback metadata) and use AAPotentialValues to identify which source values to instrument.

So far I haven't been able to do a full test run, as I'm still waiting for bank access on a suitable machine.

fodinabor committed:
Find possible callee candidates for function pointers and select among those.

This (currently) has a few restrictions:
- Attributor has to find possible callees in the first place.
- Only function pointers being handed directly to the function under investigation are handled.

fodinabor committed:
Generalize FPtr handling - now also works on memory accesses.

fodinabor committed:
Dont memoize calls across A.run().

Drop MD_dbg when stubbing a declaration.

And silence some fp info.

fodinabor committed:
Merge equivalent fp globals and stubs

ivanradanov commented on input-gen-runtimes/rt-input-gen.cpp:
```diff
@@ -576,6 +622,33 @@ void __inputgen_global(int32_t NumGlobals, VoidPtrTy Global,
+void __inputgen_access_fp(VoidPtrTy Ptr, int32_t Size, VoidPtrTy Base,
+                          VoidPtrTy *PotentialFPs, uint64_t N) {
+  if (!getInputGenRT().read<FunctionPtrTy *>(Ptr, Base, Size)) {
+    VoidPtrTy FP = PotentialFPs[rand() % N];
```
Could you try to put this in the InputGetRT class and use the rand() defined in there, would be nice to have the source of randomness be as contained (or reproducable) as possible.

ivanradanov commented on input-gen-runtimes/rt-input-gen.cpp:
```diff
@@ -375,6 +383,11 @@ struct InputGenRTTy {
     return V;
   }
 
+  template <> FunctionPtrTy *getNewValue<FunctionPtrTy *>(bool Stub, int Max) {
```
just a nitpick, maybe make have the * in FunctionPtrTy like it is in VoidPtrTy just to have things consistent

fodinabor committed:
Add requirements.txt for the scripts

fodinabor committed:
Limit the allowed AAs for the Attributor to reduce failure surface.

fodinabor committed:
Move FP selection to InputGenRTTy to contain randomness.

fodinabor committed:
Fix instrumentation dying due to eager value erasure.

fodinabor committed:
Add command line argument for disabling FP handling.

fodinabor committed:
Don't add parent function to call's callee list.

This is to prevent endless recursion via function pointers..

fodinabor committed:
Also add --disable-fp-handling to python tooling.

fodinabor committed:
Keep Attributor from replacing indirect calls with just one assumed callee by a direct call.

This makes it impossible for us to hook into this and e.g. provide our stub as an alternative..

fodinabor committed:
Fix running lit suite with fp handling.

fodinabor commented:
This is part of `input_gen` already, so closing
